typegen/src/utils.rs

```diff
@@ -131,61 +131,92 @@ fn types_equal_inner(
+        };
+
+    // We'll lazily extend our type params only if the shapes match.
+    let calc_params = || {
+        let a_params = a_parent_params.extend(&a_ty.type_params);
+        let b_params = b_parent_params.extend(&b_ty.type_params);
+        (a_params, b_params)
+    };
+    let both_composite = matches!(
+        (&a_ty.type_def, &b_ty.type_def),
+        (
+            TypeDef::Composite(_) | TypeDef::Variant(_) | TypeDef::Tuple(_),
+            TypeDef::Composite(_) | TypeDef::Variant(_) | TypeDef::Tuple(_)
+        )
+    );
     // If both IDs map to same generic param, then we'll assume equal. If they don't
     // then we need to keep checking other properties (eg Vec<bool> and Vec<u8> will have
     // different type IDs but may be the same type if the bool+u8 line up to generic params).
     if let (Some(a_idx), Some(b_idx)) = (a_generic_idx, b_generic_idx) {
-        if a_idx == b_idx {
+        if a_idx == b_idx && !both_composite {
             return true;
         }
     }
 
-    let a_ty = types.resolve(a).expect("type a should exist in registry");
-    let b_ty = types.resolve(b).expect("type b should exist in registry");
-
     // Paths differ; types won't be equal then!
-    if a_ty.path.segments != b_ty.path.segments {
-        return false;
-    }
-
-    // Names of type params don't line up, so different then!
-    if a_ty
-        .type_params
-        .iter()
-        .zip(&b_ty.type_params)
-        .any(|(a, b)| a.name != b.name)
-    {
+    if a_ty.path.segments != b_ty.path.segments && !both_composite {
         return false;
     }
 
-    // We'll lazily extend our type params only if the shapes match.
-    let calc_params = || {
-        let a_params = a_parent_params.extend(&a_ty.type_params);
-        let b_params = b_parent_params.extend(&b_ty.type_params);
-        (a_params, b_params)
-    };
+    let mut compare_fields = |a_params: &GenericsList,
+                              b_params: &GenericsList,
+                              a: &Field<PortableForm>,
+                              b: &Field<PortableForm>|
+     -> bool {
+        fn is_assoc(param: &Field<PortableForm>) -> bool {
+            param.type_name.as_ref().is_some_and(|x| x.contains("::"))
+        }
 
-    // Capture a few variables to avoid some repetition later when we recurse.
-    let mut types_equal_recurse =
-        |a: u32, a_params: &GenericsList, b: u32, b_params: &GenericsList| -> bool {
-            types_equal_inner(a, a_params, a_visited, b, b_params, b_visited, types)
+        let equal_name = a.name == b.name;
+        let a_ty_id_index = a_params.index_for_type_id(a.ty.id);
+        let b_ty_id_index = b_params.index_for_type_id(b.ty.id);
+        if !equal_name {
+            return false;
         };
 
+        match (&a.type_name, &b.type_name) {
+            (Some(_), Some(_)) if (a_ty_id_index.is_none() && b_ty_id_index.is_none()) => {
```

Comment:
would be nice with a comment why this check is needed

Reply:
added a comment